src/OrleansSQLUtils/Storage/Provider/RelationalStorageProvider.cs

```diff
@@ -0,0 +1,550 @@
+        /// <param name="version">The grain version.</param>
+        /// <param name="normalizedGrainType">Grain type without generics information.</param>
+        /// <param name="grainId">The grain ID.</param>
+        /// <param name="exceptionMessage">An optional exception message information to log.</param>
+        /// <returns>A log string to be printed.</returns>
+        private string LogString(string operationProlog, string serviceId, string providerName, string version, string normalizedGrainType, string grainId, string exceptionMessage = null)
+        {
+            const string Exception = " Exception=";
+            return $"{operationProlog}: ServiceId={serviceId} ProviderName={providerName} GrainType={normalizedGrainType} GrainId={grainId} ETag={version}{(exceptionMessage != null ? Exception + exceptionMessage : string.Empty)}.";
+        }
+
+
+        /// <summary>
+        /// Extracts the class base type without generic type information.
+        /// </summary>
+        /// <param name="classType">The class type information.</param>
+        /// <returns>The base type without generic type information.</returns>
+        /// <remarks>This is a work-around as Orleans gives grain type information in <see cref="ReadStateAsync(string, GrainReference, IGrainState)"/>
+        /// without generic type information and in <see cref="WriteStateAsync(string, GrainReference, IGrainState)"/> with it. This can be removed
+        /// when this is normalized.</remarks>
+        private static string ExtractClassBaseType(string classType)
+        {
+            //According to the type info format at https://msdn.microsoft.com/en-us/library/w3f99sx1.aspx,
+            //the generic information will always start with a back tick.
+            int indexOfGenericTypeInformation = classType.IndexOf('`');
+            if(indexOfGenericTypeInformation > -1)
+            {
+                return classType.Substring(0, indexOfGenericTypeInformation);
+            }
+
+            return classType;
+        }
+
+
+        /// <summary>
+        /// Extracts a grain ID as a string and appends the key extension with '#' infix is present.
+        /// </summary>
+        /// <param name="grainReference">The reference from which to extract the ID.</param>
+        /// <returns>The grain ID as a string.</returns>
+        /// <remarks>This likely should exist in Orleans core in more optimized form.</remarks>
+        private static string GrainIdAndExtensionAsString(GrainReference grainReference)
+        {
+            //Kudos for https://github.com/tsibelman for the algorithm. See more at https://github.com/dotnet/orleans/issues/1905.
+            object primaryKey = null;
+            string keyExtension = null;
+            try
+            {
+                primaryKey = grainReference.GetPrimaryKeyLong(out keyExtension);
+                if((long)primaryKey == 0)
+                {
+                    primaryKey = keyExtension;
```

Comment:
Thanks! I'll provide a bigger patch sometime soon addressing these things (not rebasing).